lib/source-code/source-code.js

```diff
@@ -972,6 +1051,91 @@ class SourceCode extends TokenStore {
 
     }
 
+    /**
+     * Traverse the source code and return the steps that were taken.
+     * @returns {Array<TraversalStep>} The steps that were taken while traversing the source code.
+     */
+    traverse() {
+
+        // Because the AST doesn't mutate, we can cache the steps
+        if (this.#steps) {
+            return this.#steps;
+        }
+
+        const steps = this.#steps = [];
+
+        /*
+         * This logic works for any AST, not just ESTree. Because ESLint has allowed
+         * custom parsers to return any AST, we need to ensure that the traversal
+         * logic works for any AST.
+         */
+        const emitter = createEmitter();
+        let analyzer = {
+            enterNode(node, parent) {
+                steps.push(new TraversalStep({
+                    type: "visit",
+                    target: node,
+                    phase: 1,
+                    args: [node, parent]
+                }));
+            },
+            leaveNode(node, parent) {
+                steps.push(new TraversalStep({
+                    type: "visit",
+                    target: node,
+                    phase: 2,
+                    args: [node, parent]
+                }));
```

Comment:
Questions:

1. Do these `enterNode` and `leaveNode` ever get called with the `parent` argument?
2. Are we using `args` property of `type: "visit"` steps anywhere?

Reply:
Good questions. I need to look into it.

Reply:
The answers are:

1. No, they don't, so I'll update that.
2. We aren't using `args` right now because everything is going through `NodeEventGenerator`, but I'd still like to keep these here as the foundation for the rewrite.